### DISCUSSION TITLE
adding printer entry by invoking REST call or mongo DB entry (i.e. unattended setup)

### DISCUSSION THREAD

**Discussion by artynet** [OctoFarm/OctoFarm#554] [General]
any chance to perform an unattended setup of a printer by running a python REST call or mongoDB entry ?

Regards

**Comment by davidzwa** [comment]
We dont have documentation for the API, but this is our plan in future. 
At this point the API is not validated strictly and the properties required for adding a printer are high in count, so I dont suggest using the API unless you know what you're doing and you are able to fix mistakes. I would never advise to enter database models yourself, that's even more prone to problems.
Please just use the UI, or ask for a better API which suits your need (be specific and ask for what your user case is). This API is highly impractical for your goal right now.

So with that warning out of the way, lets see about your question:
I can advise using Postman for this and then it is a question of what to put in the API request body.
I can also advise to disable login, or you have to make sure you login to the API first with a different request. This will set a session login cookie automatically, so its not hard, just an extra step: your challenge, your choice. If you want remember_me add that to the body as well.
The login route is here:
https://github.com/OctoFarm/OctoFarm/blob/e99cabd3f27f3ab6e21c27f0da5e8e319c4a2751/server_src/routes/users.js#L134
`http://OCTOFARM_SERVER:PORT/users/login` with the method POST.
> By default, LocalStrategy expects to find credentials in parameters named username and password. If your site prefers to name these fields differently, options are available to change the defaults.

Then about adding a printer:
The URL would be `OCTOFARM_SERVER:PORT/printers/add` and the method is POST.

Find that API here, although it doesnt tell you much, just yet. Read below.
https://github.com/OctoFarm/OctoFarm/blob/e99cabd3f27f3ab6e21c27f0da5e8e319c4a2751/server_src/routes/printers.js#L24
This is the model and also the parameters for the request. As you might notice you need a lot of it. This is why the API is not useful or a big challenge to call in the right way. Use with care. I repeat: if you have a specific wish, ask but be specific and detailed. Maybe we can make this a lot easier for you and others than the current situation: it will 100% prevent a lot of problems.
https://github.com/OctoFarm/OctoFarm/blob/e99cabd3f27f3ab6e21c27f0da5e8e319c4a2751/server_src/models/Printer.js#L6

```

**Reply by NotExpectedYet** [reply]
@artynet You don't need all of that modal at all btw. If you look at the printers management page the constructor I use for the API endpoint that david mentioned I've pasted below. 

https://github.com/OctoFarm/OctoFarm/blob/e99cabd3f27f3ab6e21c27f0da5e8e319c4a2751/client_src/js/printerManagementRunner.js#L2340

**Reply by NotExpectedYet** [reply]
```
  constructor(printerURL, camURL, apikey, group, name) {
    this.settingsAppearance = {
      color: "default",
      colorTransparent: false,
      defaultLanguage: "_default",
      name,
      showFahrenheitAlso: false,
    };
    this.printerURL = printerURL;
    this.camURL = camURL;
    this.apikey = apikey;
    this.group = group;
  }
}
``` 

This as an object would need posting to the end point and it will scan and add your printer.
